cli/internal/run/run.go

```diff
@@ -473,48 +476,136 @@ func filterSinglePackageGraphForDisplay(originalGraph *dag.AcyclicGraph) *dag.Ac
-		for _, dependency := range taskDefinition.TopologicalDependencies {
-			topoDeps.Add(dependency)
-		}
+
 		engine.AddTask(&core.Task{
-			Name:     taskName,
-			TopoDeps: topoDeps,
-			Deps:     deps,
+			Name:       taskName,
+			TopoDeps:   util.SetFromStrings(taskDefinition.TopologicalDependencies),
+			Deps:       deps,
+			Persistent: taskDefinition.Persistent,
 		})
 	}
 
-	if err := engine.Prepare(&core.EngineBuildingOptions{
+	engineOptions := &core.EngineBuildingOptions{
 		Packages:  rs.FilteredPkgs.UnsafeListOfStrings(),
 		TaskNames: rs.Targets,
 		TasksOnly: rs.Opts.runOpts.only,
-	}); err != nil {
+	}
+
+	if err := engine.Prepare(engineOptions); err != nil {
 		return nil, err
 	}
 
+	// Check for cycles in the DAG.
 	if err := util.ValidateGraph(engine.TaskGraph); err != nil {
 		return nil, fmt.Errorf("Invalid task dependency graph:\n%v", err)
 	}
 
+	errorThing := validatePersistentDependencies(engine, g)
+	if errorThing.PackageName != "" {
+		return nil, fmt.Errorf(
+			"Invalid persistent task dependency: %s#%s is persistent, %s#%s cannot depend on it",
+			errorThing.FailedPersistentPackageName,
+			errorThing.FailedPersistentTaskName,
+			errorThing.PackageName,
+			errorThing.TaskName,
+		)
+	}
+
 	return engine, nil
 }
 
+func validatePersistentDependencies(engine *core.Engine, graph *completeGraph) ErrorThingy {
+	// Check if the Task has persistent deps and if those deps are actually implemented
+	var persistentDepAndImplemented ErrorThingy
+
+	// TODO: Check errors returned by TaskGraph.Walk(), to make sure it didn't fail for some reason.
+	engine.TaskGraph.Walk(func(v dag.Vertex) error {
+		vertexName := dag.VertexName(v) // vertextName is a taskID
+
+		// No need to check the root node if that's where we are.
+		if strings.Contains(vertexName, core.ROOT_NODE_NAME) {
+			return nil
+		}
+
+		currentPackageName, currentTaskName := util.GetPackageTaskFromId(vertexName)
+
+		// For each "downEdge" (i.e. each task that _this_ task dependsOn)
+		// check if the downEdge is a Persistent task, and if it actually has the script implemented
+		// in that package's package.json
+		for dep := range engine.TaskGraph.DownEdges(vertexName) {
+			depTaskID := dep.(string)
+			// No need to check the root node
+			if strings.Contains(depTaskID, core.ROOT_NODE_NAME) {
+				return nil
+			}
+
+			// Parse the taskID of this dependency task
+			packageName, taskName := util.GetPackageTaskFromId(depTaskID)
+
+			// Check if the taskDefinition is Persistent
+			taskDef, ok2 := engine.GetTaskDefinition(packageName, taskName, depTaskID)
+
+			if ok2 != nil {
+				return fmt.Errorf("Cannot find task definition for %v in package %v", depTaskID, packageName)
+			}
+
+			// Check if the script is implemented
+			pkg, ok := graph.PackageInfos[packageName]
+			if !ok {
+				return fmt.Errorf("Cannot find package %v", packageName)
+			}
+			_, scriptOk := pkg.Scripts[taskName]
+			hasScript := false
+			if scriptOk {
+				hasScript = true
+			}
```

Comment:
I think this can be simplified
```suggestion
			_, hasScript := pkg.Scripts[taskName]
```